File: client/src/server/dataextractor.ts
import { LatLng } from "leaflet";
import { Ammo, Contact, GeneralSettings, Offset, Radio, TACAN } from "../@types/unit";

export class DataExtractor {
    #seekPosition = 0;
    #dataview: DataView;
    #decoder: TextDecoder;
    #buffer: ArrayBuffer;

    constructor(buffer: ArrayBuffer) {
        this.#buffer = buffer;
        this.#dataview = new DataView(this.#buffer);
        this.#decoder = new TextDecoder("utf-8");
    }

    setSeekPosition(seekPosition: number) {
        this.#seekPosition = seekPosition;
    }

    getSeekPosition() {
        return this.#seekPosition;
    }

    extractBool() {
        const value = this.#dataview.getUint8(this.#seekPosition); 
        this.#seekPosition += 1;
        return value > 0;
    }

    extractUInt8() {
        const value = this.#dataview.getUint8(this.#seekPosition); 
        this.#seekPosition += 1;
        return value;
    }

    extractUInt16() {
        const value = this.#dataview.getUint16(this.#seekPosition, true); 
        this.#seekPosition += 2;
        return value;
    }

    extractUInt32() {
        const value = this.#dataview.getUint32(this.#seekPosition, true); 
        this.#seekPosition += 4;
        return value;
    }

    extractUInt64() {
        const value = this.#dataview.getBigUint64(this.#seekPosition, true); 
        this.#seekPosition += 8;
        return value;
    }

    extractFloat64() {
        const value = this.#dataview.getFloat64(this.#seekPosition, true); 
        this.#seekPosition += 8;
        return value;
    }

    extractLatLng() {
        return new LatLng(this.extractFloat64(), this.extractFloat64(), this.extractFloat64())
    }

    extractFromBitmask(bitmask: number, position: number) {
        return ((bitmask >> position) & 1) > 0;
    }

    extractString(length?: number) {
        if (length === undefined)
            length = this.extractUInt16()
        var stringBuffer = this.#buffer.slice(this.#seekPosition, this.#seekPosition + length);
        var view = new Int8Array(stringBuffer);
        var stringLength = length;
        view.every((value: number, idx: number) => { 
            if (value === 0) {
                stringLength = idx;
                return false;
            } else 
                return true;
            });
        const value = this.#decoder.decode(stringBuffer);
        this.#seekPosition += length;
        return value.substring(0, stringLength).trim();
    }

    extractChar() {
        return this.extractString(1);
    }

    extractTACAN() {
        const value: TACAN = {
                isOn: this.extractBool(),
                channel: this.extractUInt8(),
                XY: this.extractChar(),
                callsign: this.extractString(4)
            }
        return value;
    }

    extractRadio() {
        const value: Radio = {
            frequency: this.extractUInt32(),
            callsign: this.extractUInt8(),
            callsignNumber: this.extractUInt8()
        }
        return value;
    }

    extractGeneralSettings() {
        const value: GeneralSettings = {
            prohibitJettison: this.extractBool(),
            prohibitAA: this.extractBool(),
            prohibitAG: this.extractBool(),
            prohibitAfterburner: this.extractBool(),
            prohibitAirWpn: this.extractBool(),
        }
        return value;
    }

    extractAmmo() {
        const value: Ammo[] = [];
        const size = this.extractUInt16();
        for (let idx = 0; idx < size; idx++) {
            value.push({
                quantity: this.extractUInt16(), 
                name: this.extractString(33),
                guidance: this.extractUInt8(),
                category: this.extractUInt8(),
                missileCategory: this.extractUInt8()
            });
        }
        return value;
    }

    extractContacts(){
        const value: Contact[] = [];
        const size = this.extractUInt16();
        for (let idx = 0; idx < size; idx++) {
            value.push({
                ID: this.extractUInt32(), 
                detectionMethod: this.extractUInt8()
            });
        }
        return value;
    }

    extractActivePath() {
        const value: LatLng[] = [];
        const size = this.extractUInt16();
        for (let idx = 0; idx < size; idx++) {
            value.push(this.extractLatLng());
        }
        return value;
    }

    extractOffset() {
        const value: Offset = {
            x: this.extractFloat64(),
            y: this.extractFloat64(),
            z: this.extractFloat64(),
        }
        return value;
    }
}

File: scripts/python/API/data_extractor.py
import struct
from typing import List
from data_types import LatLng, TACAN, Radio, GeneralSettings, Ammo, Contact, Offset

class DataExtractor:
    def __init__(self, buffer: bytes):
        self._seek_position = 0
        self._buffer = buffer
        self._length = len(buffer)
    
    def set_seek_position(self, seek_position: int):
        self._seek_position = seek_position
    
    def get_seek_position(self) -> int:
        return self._seek_position
    
    def extract_bool(self) -> bool:
        value = struct.unpack_from('<B', self._buffer, self._seek_position)[0]
        self._seek_position += 1
        return value > 0
    
    def extract_uint8(self) -> int:
        value = struct.unpack_from('<B', self._buffer, self._seek_position)[0]
        self._seek_position += 1
        return value
    
    def extract_uint16(self) -> int:
        value = struct.unpack_from('<H', self._buffer, self._seek_position)[0]
        self._seek_position += 2
        return value
    
    def extract_uint32(self) -> int:
        value = struct.unpack_from('<I', self._buffer, self._seek_position)[0]
        self._seek_position += 4
        return value
    
    def extract_uint64(self) -> int:
        value = struct.unpack_from('<Q', self._buffer, self._seek_position)[0]
        self._seek_position += 8
        return value
    
    def extract_float64(self) -> float:
        value = struct.unpack_from('<d', self._buffer, self._seek_position)[0]
        self._seek_position += 8
        return value
    
    def extract_lat_lng(self) -> LatLng:
        lat = self.extract_float64()
        lng = self.extract_float64()
        alt = self.extract_float64()
        return LatLng(lat, lng, alt)
    
    def extract_from_bitmask(self, bitmask: int, position: int) -> bool:
        return ((bitmask >> position) & 1) > 0
    
    def extract_string(self, length: int = None) -> str:
        if length is None:
            length = self.extract_uint16()
        
        string_buffer = self._buffer[self._seek_position:self._seek_position + length]
        
        # Find null terminator
        string_length = length
        for idx, byte_val in enumerate(string_buffer):
            if byte_val == 0:
                string_length = idx
                break
        
        try:
            value = string_buffer[:string_length].decode('utf-8').strip()
        except UnicodeDecodeError:
            value = string_buffer[:string_length].decode('utf-8', errors='ignore').strip()
        
        self._seek_position += length
        return value
    
    def extract_char(self) -> str:
        return self.extract_string(1)
    
    def extract_tacan(self) -> TACAN:
        return TACAN(
            is_on=self.extract_bool(),
            channel=self.extract_uint8(),
            xy=self.extract_char(),
            callsign=self.extract_string(4)
        )
    
    def extract_radio(self) -> Radio:
        return Radio(
            frequency=self.extract_uint32(),
            callsign=self.extract_uint8(),
            callsign_number=self.extract_uint8()
        )
    
    def extract_general_settings(self) -> GeneralSettings:
        return GeneralSettings(
            prohibit_jettison=self.extract_bool(),
            prohibit_aa=self.extract_bool(),
            prohibit_ag=self.extract_bool(),
            prohibit_afterburner=self.extract_bool(),
            prohibit_air_wpn=self.extract_bool()
        )
    
    def extract_ammo(self) -> List[Ammo]:
        value = []
        size = self.extract_uint16()
        for _ in range(size):
            value.append(Ammo(
                quantity=self.extract_uint16(),
                name=self.extract_string(33),
                guidance=self.extract_uint8(),
                category=self.extract_uint8(),
                missile_category=self.extract_uint8()
            ))
        return value
    
    def extract_contacts(self) -> List[Contact]:
        value = []
        size = self.extract_uint16()
        for _ in range(size):
            value.append(Contact(
                id=self.extract_uint32(),
                detection_method=self.extract_uint8()
            ))
        return value
    
    def extract_active_path(self) -> List[LatLng]:
        value = []
        size = self.extract_uint16()
        for _ in range(size):
            value.append(self.extract_lat_lng())
        return value
    
    def extract_offset(self) -> Offset:
        return Offset(
            x=self.extract_float64(),
            y=self.extract_float64(),
            z=self.extract_float64()
        )